### ISSUE TITLE
Unclear behavior for enhancers in certain edge cases

### ISSUE THREAD

**Issue by ganemone** [fusionjs/fusionjs#573]
After investigating #572 I realized we have some undefined behavior surrounding certain edge cases for enhancers. The following edge cases have the following behavior:

**Enhancer on unregistered token with no dependencies**
  - does not throw at resolution time
  - does not execute enhancer
  
Example:
```js
  const app = new App();
  app.enhance(SomeEnhancer, TokenA);
  return app;
```

**Enhancer on unregistered token with optional dependency**
  - throws at resolution time

Example:
```js
  const app = new App();
  app.enhance(SomeEnhancer, TokenA);
  app.register(createPlugin({
    deps: {
      a: TokenA.optional
    }
  }));
  return app;
```

I think our strategy for DI logic should be focused on correctness, eliminating dead code, and validating intent as much as possible. This is why we have features such as disallowing orphan configuration registrations. The only use case I can think of for allowing enhancers to be registered on unregistered tokens would be for providing a default implementation. For example:

```js
app.enhance(TokenA, (a) => {
  if (!a) {
    return 'some-default';
  }
  return a;
});
```

However, I think mixing semantics of enhancing existing registrations and providing default implementations is not a good idea. If we want to support enhancing optional tokens, we can do so by extending the `app.enhance` semantics to support it:

```js
app.enhance(TokenA.optional, (a) => {
  if (!a) {
    return 'some-default';
  }
  return a;
});
```

Alternatively, we could introduce a new API specifically designed for providing default implementations (which we have discussed and rejected in the past).

```js
app.default(TokenA, 'some-default');
```

For now, my recommendation would be to fix #572 and to fail dependency resolution if an enhancer is registered on a Token with no implementation. If we have a need in the future for providing default implementations on Tokens, we can explore options listed above.

**Comment by rtsao** [comment]
I think one drawback to the notion of "default" implementations is it leads to dead code if the default is replaced with something else. Are there specific use cases where this would be really helpful?

Obviously I could imagine it being more convenient, but this essentially encourages dead code while also increasing the complexity of understanding how/which values become registered at a specific token.

**Comment by ganemone** [comment]
I agree about the dead code argument. I haven't seen any specific use cases where this would be needed yet, so I think it is fair to not implement anything surrounding defaults unless a compelling use case arises.
